tests/backend/checkmach/test_signatures_separate_b.ml

```diff
@@ -0,0 +1,80 @@
+(* Tests showing how [zero_alloc] information from a signature can be used in
+   checking uses of the corresponding functions.
+
+   This is part of a test of separate compilation. This file uses the library
+   [test_signatures_separate_a.ml]. The dune rules delete its compiled
+   implementation first, so we're only using its signature.
+*)
+
+module A = Test_signatures_separate_a
+
+(* Basic behavior with strict and assume *)
+let[@zero_alloc] f_id_is_za x = A.f_id x
+
+let[@zero_alloc strict] f_id_is_not_za_strict x = A.f_id x
+
+let[@zero_alloc] f_id_strict_is_za x = A.f_id_strict x
+
+let[@zero_alloc strict] f_id_strict_is_za_strict x = A.f_id_strict x
+
+let[@zero_alloc] f_tuple_assume_is_za x = A.f_tuple_assume x
+
+let[@zero_alloc strict] f_tuple_assume_is_not_za_strict x = A.f_tuple_assume x
+
+let[@zero_alloc] f_tuple_no_assume_is_not_za x = A.f_tuple_no_assume x
+
+let[@zero_alloc strict] f_tuple_no_assume_is_not_za_strict x =
+  A.f_tuple_no_assume x
+
+(* opt behavior - these functions are marked zero-alloc opt in A. We use expect
+   to accept these as functions as "zero_alloc opt" in B always (because when
+   compiling with `-zero_alloc_check all` we make those assumptions available,
+   and when compiling without it we don't check "zero_alloc opt" facts here.
+
+   However, we can also use these assumptions even to prove normal non-opt
+   zero_alloc facts, when compiling this module with -zero-alloc-check all. This
+   is basically a bug, but because the whole build is compiled with one
+   -zero-alloc-check X setting, we're not too worried about it.  These tests
+   document the current behavior, but it would be OK if we changed things so
+   that the ones without "opt" fail.
```

Comment:
I am still confused about this again, sorry. If we compile `a.ml` with `-zero-alloc-check default`, is the fact that `f_id_opt` is annotated with "zero_alloc opt" recorded in the `a.cmi`? If so, why does it need to be recorded? 

If it's not recorded, then compiling `b.ml` with `-zero-alloc-check all` (assuming that  `a.cmx` is not available) will fail to prove that `foo` is zero_alloc:
```
let[@zero_alloc] foo x = A.f_id_opt x
```
Would it address the bug you are referring to in this comment? Is there a problem with this approach?

Reply:
> I am still confused about this again, sorry. If we compile `a.ml` with `-zero-alloc-check default`, is the fact that `f_id_opt` is annotated with "zero_alloc opt" recorded in the `a.cmi`? If so, why does it need to be recorded?

Yes, it is.  I think there's no particular reason this needs to be the case to make zero_alloc work.  But I think it's the right thing to do for three reasons:
- I don't think this would be the right semantics.  `@zero_alloc opt` means "I am zero alloc if the whole build is compiled with optimization" (or, really, with check all).  The meaning of the attribute already has an "if" in it - we don't need to delete it if the condition in the if is false.
- I don't think changing this would fix the bug (more on that below)
- I'm very wary of making typechecking change based on optimization level. Deleting information from a module type depending on what optimization flags we pass would be an unusual and error-prone thing to do.

> If it's not recorded, then compiling `b.ml` with `-zero-alloc-check all` (assuming that `a.cmx` is not available) will fail to prove that `foo` is zero_alloc:
> 
> ```
> let[@zero_alloc] foo x = A.f_id_opt x
> ```
> 
> Would it address the bug you are referring to in this comment? Is there a problem with this approach?

I don't think this is right.  We already fail to prove this fact when compiling both modules without `-zero-alloc-check all`.  The "bug" is about what happens when we compile both modules _with_ this flag. In that case, we will successfully prove this fact, and that's a little weird because the attribute doesn't have "opt" in it.

I have tried to clarify the comment a little.